tests/type-checks.ts

```diff
@@ -86,9 +93,263 @@ async function closureSoICanUseAwait() {
     assert<Has<typeof b, () => Promise<number>>>(true);
     assert<IsAny<typeof b>>(false);
     const subproxy = proxy.c;
-    assert<Has<typeof subproxy, { d: Promise<number> }>>(true);
+    assert<Has<typeof subproxy, Promise<{ d: number }>>>(true);
     assert<IsAny<typeof subproxy>>(false);
     const copy = await proxy.c;
     assert<Has<typeof copy, { d: number }>>(true);
   }
+
+  {
+    Comlink.wrap(new MessageChannel().port1);
+    Comlink.expose({}, new MessageChannel().port2);
+    const connection = new RTCPeerConnection();
+    const channel = connection.createDataChannel("comlink");
+
+    interface Baz {
+      baz: number;
+      method(): number;
+    }
+
+    class Foo {
+      constructor(cParam: string) {
+        const self = this;
+        assert<IsExact<typeof self.proxyProp, Bar & Comlink.ProxyMarked>>(true);
+      }
+      prop1: string = "abc";
+      proxyProp = Comlink.proxy(new Bar());
+      methodWithTupleParams(...args: [string] | [number, string]): number {
+        return 123;
+      }
+      methodWithProxiedReturnValue(): Baz & Comlink.ProxyMarked {
+        return Comlink.proxy({ baz: 123, method: () => 123 });
+      }
+      methodWithProxyParameter(param: Baz & Comlink.ProxyMarked): void {}
+    }
+
+    class Bar {
+      prop2: string | number = "abc";
+      method(param: string): number {
+        return 123;
+      }
+      methodWithProxiedReturnValue(): Baz & Comlink.ProxyMarked {
+        return Comlink.proxy({ baz: 123, method: () => 123 });
+      }
+    }
+    const proxy = Comlink.wrap<Foo>(Comlink.windowEndpoint(self));
+    assert<IsExact<typeof proxy, Comlink.Remote<Foo>>>(true);
+
+    proxy[Comlink.releaseProxy]();
+    const endp = proxy[Comlink.createEndpoint]();
+    assert<IsExact<typeof endp, Promise<MessagePort>>>(true);
+
+    assert<IsAny<typeof proxy.prop1>>(false);
+    assert<Has<typeof proxy.prop1, Promise<string>>>(true);
+
+    const r1 = proxy.methodWithTupleParams(123, "abc");
+    assert<IsExact<typeof r1, Promise<number>>>(true);
+
+    const r2 = proxy.methodWithTupleParams("abc");
+    assert<IsExact<typeof r2, Promise<number>>>(true);
+
+    assert<
+      IsExact<typeof proxy.proxyProp, Comlink.Remote<Bar & Comlink.ProxyMarked>>
```

Comment:
Technically, the remote side is never `ProxyMarked`, as the symbol doesn’t get transferred. The typings should probably get adjusted for that.

Reply:
You are correct that the symbol property cannot be accessed from this side, but what the `& Comlink.ProxyMarked` is saying here is that it's proxy marked _on the remote side_. This is then the input to `Comlink.Remote<T>`, which needs to look at whether it is proxy marked to decide whether to clone or proxy. So the `& Comlink.ProxyMarked` is correct here, even if it may be confusing at first (but users won't usually have to write this type, this is just for the tests).

Now one thing I want to mention for honesty is that despite what what I said above, `Comlink.Remote<T>` does **not** remove the `[proxyMarker]: boolean` from the type, but treats it like a regular property and converts it to `[proxyMarker]: Promise<boolean>`. Of course, as you said, this cannot actually be accessed at runtime because symbols can't be cloned (although comlink could theoretically detect it as a well-known symbol and use a transfer handler, but there is no use case for this).
The reason for this is that the inverse type, `Local<T>`, needs to check for proxy marking too. It does so in this definition:

```ts
export type UnproxyOrClone<T> = T extends RemoteObject<ProxyMarked> // This is essentially { [proxyMarked]: Promise<true> }
  ? Local<T>
  : T;
```

If the `proxyMarked` was removed, this would not work and we would need to add some other kind of type-only "marker", but I this is the more natural solution imo. It is also the easiest to maintain because it will "just work" even if `Remote<T>` is extended to handle more types in the future. It makes the types nicely symmetric ✨

Reply:
Yeah, I see where you are coming from with `Local<T>`. Let’s leave it in for now and see if it causes any problems for anyone :)